src/loaders/common/loader_helper.py

```diff
@@ -289,6 +292,41 @@ def start_podman_service(uid: int):
     return proc
 
 
+def get_containers_config():
+    """Get containers.conf file at home directory."""
+    home = os.path.expanduser("~")
+    conf_path = os.path.join(home, CONTAINERS_CONF_PATH)
+    config = configparser.ConfigParser()
+    config.read(conf_path)
+    return config, conf_path
+
+
+def is_config_modification_required():
+    """check if the config requires modification."""
+    config, _ = get_containers_config()
+    if not config.has_section("containers"):
+        return True
+    for key, val in CONTAINERS_CONF_PARAMS.items():
+        if config.get("containers", key, fallback=None) != val:
+            return True
+    return False
+
+
+def setup_callback_server_logs():
+    """Set up containers.conf file for the callback server logs."""
+    config, conf_path = get_containers_config()
+    if not config.has_section("containers"):
+        config.add_section("containers")
+
+    for key, val in CONTAINERS_CONF_PARAMS.items():
+        if config.get("containers", key, fallback=None) != val:
+            config.set("containers", key, val)
+
+    with open(conf_path, 'w') as configfile:
```

Comment:
There's a potential race condition here where if another process writes to the file between the read and write in this function the other process's writes will be clobbered. Can you see if there's a way to lock the file or some other way to resolve the race condition? Don't spend a ton of time on this as it seems pretty improbable... maybe no more than an hour?

Reply:
I cannot access containers.conf at your home dir. You cannot access mine either.

Reply:
The process could also be owned by the user

Reply:
Updated!

Reply:
That's a good start, but I'd use an `RLock` rather than plain `Lock`, and unless I'm misunderstanding, you need to create a global `RLock` object and use that in the method, otherwise every time the method is called a new unlocked lock object will be created and so the call will never block

Reply:
Unfortunately using a thread lock will only protect the file from overwrites from this specific script. What we want here is some sort of OS-level lock.

Reply:
https://docs.python.org/3/library/fcntl.html#fcntl.flock

Reply:
Updated!

Reply:
The current implementation now locks the file for all other processes, but it still allows for a race condition because the flock is only around the write, not the read:

* Process A reads the file
* Process B writes to the file
* Process A locks the file
* Process A writes to the file, clobbering Process B's changes
* Process A releases the lock

The flock needs to be created *before* the process reads the file

Reply:
👍

Reply:
Getting closer, but I would only lock the file after the config check and user input. The key part where the lock is required is the read, modify, write step *after* the user gives permission to modify the file. Then I think the locking can be abstracted into loader_helper.py and out of the main logic

Reply:
config check requires to read the file. Therefore I put it before.

Reply:
But you don't need a lock just to read and close it. You only need the lock when you do the read/modify/write step.

Reply:
Process A reads the file
Process A gets permission to modify the file
Process B writes to the file
Process A locks the file

Isn't this same as the race condition you mentioned here: https://github.com/kbase/collections/pull/426#discussion_r1435315395

Reply:
wait, are you saying have multiple locks? Like

lock
read
release lock

lock
write
release lock

Reply:
> Process A reads the file
> Process A gets permission to modify the file
> Process B writes to the file
> Process A locks the file

This is totally fine - after the lock, then `setup_callback_server_logs` will read the file again (including B's changes), modify the config, write, and unlock, so both As and Bs changes are preserved in the file.

The issue would be if process a read the file, locked, and then changed the config and wrote it, since if B wrote to the file post read but pre lock B's changes would be overwritten.

The only reason we're worrying about this, btw, is that this is a shared config file for Docker as opposed to a file that only our code is supposed to touch, like everything in /global/.../kbase/collections

Reply:
Also I think the lock can be moved back into `setup_callback_server_logs`

Reply:
👍

Reply:
Looks good. I think the last thing is to surround the lock and subsequent steps with a `try` and put the unlock in a `finally` to ensure sure the file is always unlocked

Reply:
👍

Reply:
:+1: